examples/test_driven_software_defined_assets/consumption_datamart/assets/acme_lake/sql_lake_io_manager.py

```diff
@@ -0,0 +1,68 @@
+from abc import abstractmethod, ABC
+
+from dagster import IOManager, io_manager
+from dagster_pandas import DataFrame
+
+
+class SQLLakeIOManager(IOManager, ABC):
+
+    def __init__(self, datawarehouse_resource) -> None:
+        super().__init__()
+        self.datawarehouse_resource = datawarehouse_resource
+
+    def handle_output(self, context, obj: DataFrame):
+        pass
+
+    @abstractmethod
+    def load_input(self, context):
+        pass
+
+
+class AssetSQLLakeIOManager(SQLLakeIOManager):
+
+    def load_input(self, context):
+        asset_out = context.solid_def.outs['result']
+        metadata = asset_out.metadata
+        dagster_type = asset_out.metadata['dagster_type']
+
+        df = self.datawarehouse_resource.read_sql_query(metadata['load_sql'])
+        typed_df = dagster_type.convert_dtypes(df)
+
+        dagster_type.calculate_data_quality(typed_df)
+
+        return typed_df
+
+
+class ForeignAssetSQLLakeIOManager(SQLLakeIOManager):
+
+    def load_input(self, context):
+        asset_out = context.upstream_output
+        metadata = asset_out.metadata
+        dagster_type = asset_out.metadata['dagster_type']
+
+        df = self.datawarehouse_resource.read_sql_query(metadata['load_sql'])
+        typed_df = dagster_type.convert_dtypes(df)
+
+        dagster_type.calculate_data_quality(typed_df)
+
+        # When AssetObservation is implemented, emit metadata about the (foreign) Asset just loaded
+        # yield AssetObservation(
+        #     asset_key=context.asset_key,
+        #     metadata=dagster_type.extract_event_metadata(typed_df),
+        # )
```

Comment:
@sryza Will I be able to emit AssetObservations them from a IO Manager like this?

Related, do you have a rough ETA for AssetObservations are likely to be implemented?

Reply:
Yes - we'll also likely support some sugar so that you don't need to do `asset_key=context.asset_key`. For  example, `context.log_asset_observation(metadata=...)`.

Let me get back to you on ETA. We have a meeting tomorrow morning to chat about an implementation plan.

Reply:
Thanks.  Good to know I'm heading in the right direction